python_api: let approx() take nonnumeric values

jvansanten opened the pull request:
This makes `x == approx(y)` equivalent to` x == y` if x or y are not instances of `decimal.Decimal` or `numbers.Complex`, rather than raising a `TypeError`. This makes approximate equality a strict superset of equality (for scalars, sequences, and dicts), which comes in handy e.g. in situations where dictionaries can contain optional float values.

Closes #4079.

- [X] Include documentation when adding new features.
- [X] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.
- [X] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number). See the [github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more information.
- [x] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/master/changelog/README.rst) for details.
- [x] Add yourself to `AUTHORS` in alphabetical order.

Long-time user, first-time contributor, so would appreciate advice on whether this needs an entry in the changelog or authors file. The docs do not specifically mention that there are values that compare equal but do not compare approximately equal, so I left those alone.

nicoddemus commented:
Thanks a lot @jvansanten for the PR, and sorry for the silence on this one!

To summarize for others reading this PR, it addresses the part about falling back to standard equality for non-numeric types in containers from https://github.com/pytest-dev/pytest/issues/4079:

> 1. Allow non-Number elements in sequences / mappings. For those elements, check for exact equality only.
> 2. Add support for nested sequences / mappings (e.g. [1, 2, [3, 4]] == approx([1, 2, [3, 4]]))


```python
# test_foo.py
import pytest

def test():
    x1 = {"foo": 1.0000005, "bar": None}
    x2 = {"foo": 1.0000006, "bar": None}
    assert x1 == pytest.approx(x2)

def test2():
    x1 = {"foo": 1.0000005, "c": {"bar": None}}
    x2 = {"foo": 1.0000006, "c": {"bar": None}}
    assert x1 == pytest.approx(x2)
```

**master**

```
======================== test session starts ========================
collected 2 items

.tmp\test_foo.py FF                                            [100%]

============================= FAILURES ==============================
_______________________________ test ________________________________

    def test():
        x1 = {"foo": 1.0000005, "bar": None}
        x2 = {"foo": 1.0000006, "bar": None}
>       assert x1 == pytest.approx(x2)
E       TypeError: cannot make approximate comparisons to non-numeric values: {'foo': 1.0000006, 'bar': None}  at key='bar'

.tmp\test_foo.py:6: TypeError
_______________________________ test2 _______________________________

    def test2():
        x1 = {"foo": 1.0000005, "c": {"bar": None}}
        x2 = {"foo": 1.0000006, "c": {"bar": None}}
>       assert x1 == pytest.approx(x2)
E       TypeError: pytest.approx() does not support nested dictionaries: key='c' value={'bar': None}
E         full mapping={'c': {'bar': None}, 'foo': 1.0000006}

.tmp\test_foo.py:11: TypeError
====================== short test summary info ======================
FAILED .tmp\test_foo.py::test - TypeError: cannot make approximate ...
FAILED .tmp\test_foo.py::test2 - TypeError: pytest.approx() does no...
========================= 2 failed in 0.20s =========================
```

**this PR**

```
======================== test session starts ========================
collected 2 items

.tmp\test_foo.py .F                                            [100%]

============================= FAILURES ==============================
_______________________________ test2 _______________________________

    def test2():
        x1 = {"foo": 1.0000005, "c": {"bar": None}}
        x2 = {"foo": 1.0000006, "c": {"bar": None}}
>       assert x1 == pytest.approx(x2)
E       TypeError: pytest.approx() does not support nested dictionaries: key='c' value={'bar': None}
E         full mapping={'c': {'bar': None}, 'foo': 1.0000006}

.tmp\test_foo.py:11: TypeError
====================== short test summary info ======================
FAILED .tmp\test_foo.py::test2 - TypeError: pytest.approx() does no...
==================== 1 failed, 1 passed in 0.16s ====================
```

When this gets merged, we should create a new issue to handle point 2.

nicoddemus commented on testing/python/approx.py:
```diff
@@ -466,18 +476,48 @@ def test_foo():
+            pytest.param({"key": "string"}, id="dict-with-string"),
+        ],
+    )
+    def test_nonnumeric_okay_if_equal(self, x):
```
We should include another test which also checks the message when equality fails:

```python
def test():
    x1 = {"foo": 1.0000005, "bar": None, "foobar":1}
    x2 = {"foo": 1.0000006, "bar": None}
    assert x1 == pytest.approx(x2)
```

This currently outputs:

```
_______________________________ test ________________________________

    def test():
        x1 = {"foo": 1.0000005, "bar": None, "foobar":1}
        x2 = {"foo": 1.0000006, "bar": None}
>       assert x1 == pytest.approx(x2)
E       assert {'bar': None, 'foo': 1.0000005, 'foobar': 1} == <[TypeError("bad operand type for abs(): 'NoneType'") raised in repr()] ApproxMapping object at 0x281cf9f3430>
E        +  where <[TypeError("bad operand type for abs(): 'NoneType'") raised in repr()] ApproxMapping object at 0x281cf9f3430> = <function approx at 0x00000281CECB3CA0>({'bar': None, 'foo': 1.0000006})
E        +    where <function approx at 0x00000281CECB3CA0> = pytest.approx
```

That message should be improved. 👍

jvansanten replied:
Right, I did not think to check the message. Should be fixed now.

nicoddemus commented on testing/python/approx.py:
```diff
@@ -346,6 +349,13 @@ def test_dict_wrong_len(self):
 
+    def test_dict_nonnumeric(self):
+        assert {"a": 1.0, "b": None} == pytest.approx({"a": 1.0, "b": None})
+        assert {"a": 1.0, "b": 1} != pytest.approx({"a": 1.0, "b": None})
```
Same problem here with the assertion message:

```python

    def test():
        x1 = {"foo": 1.0000005, "bar": 1}
        x2 = {"foo": 1.0000006, "bar": None}
>       assert x1 == pytest.approx(x2)
E       assert {'bar': 1, 'foo': 1.0000005} == <[TypeError("bad operand type for abs(): 'NoneType'") raised in repr()] ApproxMapping object at 0x1c5c4b13340>
E        +  where <[TypeError("bad operand type for abs(): 'NoneType'") raised in repr()] ApproxMapping object at 0x1c5c4b13340> = <function approx at 0x000001C5C3DD3CA0>({'bar': None, 'foo': 1.0000006})
E        +    where <function approx at 0x000001C5C3DD3CA0> = pytest.approx
```

nicoddemus reviewed (changes requested):
> Long-time user, first-time contributor, so would appreciate advice on whether this needs an entry in the changelog or authors file. The docs do not specifically mention that there are values that compare equal but do not compare approximately equal, so I left those alone.

Great, thanks a lot @jvansanten!

Indeed we should also add some docs:

1. Update the ``approx`` docstring with some examples of the new behavior.
2. Add a CHANGELOG entry, I think adding a small example also helps.
3. Definitely add yourself to `AUTHORS`, this is a great contribution! 😁

jvansanten commented:
Thanks for the feedback! I made `ApproxScalar.__repr__()` more robust as requested, updated the docstring, and added a changelog entry. I called it a bugfix since the feature in question [seems to have regressed in a patch release](https://github.com/pytest-dev/pytest/commit/cd2085ee718fb297537638bb33a70fc781ea450f). Since it appears that the pre-3.7.1 behavior was undefined, though, it's certainly up for debate.

nicoddemus commented:
Thanks @jvansanten!

> I called it a bugfix since the feature in question seems to have regressed in a patch release. Since it appears that the pre-3.7.1 behavior was undefined, though, it's certainly up for debate.

Indeed that makes sense, but I changed it to an improvement for two reasons:

1. As you stated in the changelog that was undocumented, making it an supported feature feels more official.
2. Even if it was a bugfix, I would prefer not to backport it to `6.0.x`, given the possibility of breaking test suites.

👍

nicoddemus reviewed (approved):
Awesome work @jvansanten, thanks!

Zac-HD reviewed (approved):
Looks good to me - thanks, @jvansanten!

@nicoddemus, is there anything else needed before we merge?

nicoddemus commented:
> @nicoddemus, is there anything else needed before we merge?

Not from my side. If nobody states otherwise, we can merge this tomorrow. 👍
